File: BingitOn.py
"""
/*	Turner Atwood
 *	10/2/18
 *	Bing it On [4.1] (https://open.kattis.com/problems/bing)
 *	Prefix Tree
 */
"""

# TOOO SLOWWWWW

class Node():
	count = 0
	children = None
	val = None

	def __init__(self, letter):
		self.count = 1
		self.val = letter
		self.children = set()

	def _find(self, match_letter):
		for potential_node in self.children:
			if potential_node.val == match_letter:
				return potential_node
		return None

class PrefixTree():
	root = None

	def __init__(self, letters):
		first_letter = letters[0]
		self.root = Node(None)
		self._add_to_tree(letters, self.root)

	# Only add once a str has been searched
	def _add_to_tree(self, letters, current_parent):
		# letters = letters.copy()
		while letters:
			letter = letters.pop(0)
			new_node = Node(letter)
			current_parent.children.add(new_node)
			current_parent = new_node

	# Return letters not found and reference node
	def _is_prefix(self, letters, do_add = True):
		current_node = self.root
		# letters = letters.copy()
		while letters:
			match_letter = letters[0]
			next_node = current_node._find(match_letter)
			if next_node:
				letters = letters[1:]
				if do_add:
					current_node.count += 1
				current_node = next_node
			else:
				break
		if do_add:
			current_node.count += 1	
		return (letters, current_node)





def main():
	num_words = int(input())
	pTree = None
	output = ""
	for _ in range(num_words):
		letters = list(input())
		if not pTree:
			pTree = PrefixTree(letters)
			output += '0\n'
		else:
			leftovers, last_node = pTree._is_prefix(letters)
			if not leftovers:
				output += f"{last_node.count-1}\n"
			else:
				pTree._add_to_tree(leftovers, last_node)
				output += '0\n'
	print(output)

if __name__ == "__main__":
	main()

File: Favorites/BingitOn.java

/*	Turner Atwood
 *	10/22/18
 *	Bing it On [4.3] (https://open.kattis.com/problems/bing)
 *	Prefix Tree - Python was too slow
 */

import java.util.HashMap;
import java.util.ArrayList;
import java.io.BufferedReader; 
import java.io.IOException; 
import java.io.InputStreamReader; 
import java.util.StringTokenizer;
import java.lang.StringBuilder;

class BingitOn {

	public static void main(String args[]) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int num = Integer.parseInt(br.readLine());
		PrefixTree pTree = new PrefixTree();

		//StringBuilder to reduce the time to print/append lots of strings
		StringBuilder sb = new StringBuilder();
		for (int i = 0; i < num; i++) {
			String line = br.readLine();
			sb.append(pTree.isPrefix(line)+"\n");
		}

		System.out.print(sb);
	}

	static class PrefixTree {
		Node root;

		PrefixTree() {
			root = new Node(' ');
		}

		// Add a line of nodes based on the given string
		//	 starting at the given node
		void addToTree(String letters, Node current_parent) {
			for (int i = 0; i < letters.length(); i++) {
				char new_letter = letters.charAt(i);
				Node new_Node = new Node(new_letter);
				current_parent.children.put(new_letter, new_Node);
				current_parent = new_Node;
			}
		}

		// Returns how many previous strings use this string as a prefix
		//	Also adds to the tree if a string is not a prefix and updates counters
		int isPrefix(String letters) {
			Node current_node = root;
			for (int i = 0; i < letters.length(); i++) {
				char match_letter = letters.charAt(i);
				Node next_node = current_node.find(match_letter);
				
				// Not a prefix
				if (next_node == null) {
					addToTree(letters.substring(i), current_node);
					return 0;
				}
				next_node.count++;
				current_node = next_node;
			}
			return current_node.count;
		}
	}

	// NODE SUBCLASS
	static class Node {
		int count;
		char val;
		HashMap<Character,Node> children;

		Node(char in_val) {
			val = in_val;
			count = 0;
			children = new HashMap<Character,Node>();
		}

		// Find the value for the given character in children
		Node find(char match_letter) {
			if (children.containsKey(match_letter)) {
				return children.get(match_letter);
			}
			return null;
		}
	}
}